source/MaterialXContrib/MaterialXNode/MaterialXNode.cpp

```diff
@@ -120,59 +120,66 @@ bool MaterialXNode::getInternalValue(const MPlug& plug, MDataHandle& dataHandle)
-        try
+        if (_documentFilePath.length() == 0)
+        {
+            return;
+        }
+
+        if (!_document || reloadDocument)
         {
-            if (_documentFilePath.length() == 0)
-            {
-                return;
-            }
-
-            if (!_document)
-            {
-                _document = MaterialXMaya::loadDocument(
-                    _documentFilePath.asChar(), Plugin::instance().getLibrarySearchPath()
-                );
-            };
-
-            if (_elementPath.length() == 0)
-            {
-                // When an empty element path is passed to MaterialXData's
-                // constructor, the first renderable element is selected, which is
-                // a handy feature when creating the node with the command.
-                // However this automatic behavior would complicate state
-                // transitions on attribute edits after the node has been created.
-                // So if the element path attribute is empty, bail early and don't
-                // attempt to create a MaterialXData.
-                //
-                return;
-            }
-
-            _materialXData.reset(new MaterialXData(
-                _document,
-                _elementPath.asChar(),
-                Plugin::instance().getLibrarySearchPath()
-            ));
-
-            MaterialXMaya::registerFragment(
-                _materialXData->getFragmentName(), _materialXData->getFragmentSource()
+            _document = MaterialXMaya::loadDocument(
+                _documentFilePath.asChar(), Plugin::instance().getLibrarySearchPath()
             );
+        };
+
+        if (_elementPath.length() == 0)
+        {
+            // When an empty element path is passed to MaterialXData's
+            // constructor, the first renderable element is selected, which is
+            // a handy feature when creating the node with the command.
+            // However this automatic behavior would complicate state
+            // transitions on attribute edits after the node has been created.
+            // So if the element path attribute is empty, bail early and don't
+            // attempt to create a MaterialXData.
+            //
+            return;
         }
-        catch (std::exception& e)
+
+        _materialXData.reset(new MaterialXData(
+            _document,
+            _elementPath.asChar(),
+            Plugin::instance().getLibrarySearchPath()
+        ));
+
+        MaterialXMaya::registerFragment(
+            _materialXData->getFragmentName(), _materialXData->getFragmentSource()
+        );
+
+        if (reloadDocument)
         {
-            MString message("Failed to create shader for ");
-            message += typeName();
-            message += " '";
-            message += name();
-            message += "': ";
-            message += e.what();
-            MGlobal::displayError(message);
+            // TODO: Figure out a better way to refresh the viewport
```

Comment:
Until we figure out a better way, this is a workaround for refreshing the viewport.

Reply:
For Arnold, how will it know that something is changed. Our assumption is that the node attributes will somehow be set dirty.

Reply:
It's a good question, maybe something for @ppenenko to investigate next?

Reply:
Not sure I understand. Shouldn't Arnold always read the latest contents of the document file when it's time to render? Or is it expected to receive some kind of live updates? Either way, can we consider this in a future PR? This one improves things on the VP2 side of things anyway.

Reply:
I think the point is that Arnold needs to know when to render things. It needs to know when things have changed. But ya, I think it's for a future review to deal with.

Reply:
If it's a one-time update then maybe we could call an Arnold-specific command from this one?

Reply:
There are no Arnold specific commands. It just monitors for node changes. Isn't it as simple as getting a plug and setting it dirty -- maybe a 1 liner?

Reply:
The only such mechanism I'm aware of is the dgdirty command, but I'm not sure whether it will do what you want or not. Is there a way for me to get a build with the Arnold changes to try things out?

Reply:
Yep, I think it's important to know the exact mechanism Arnold would use to monitor for node changes. An attribute value changed callback perhaps? On the other hand, since we know all the situations when the document is reloaded on our side, we could just call whatever command or callback Arnold gives us without the need to monitor from their side.

Reply:
Makes sense. Please ping Krishnan for what they monitor and get this in first.